daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/dpath/NodeInfo.scala

```diff
@@ -382,18 +380,28 @@ object NodeInfo extends Enum {
    * All complex types are represented by this one type object.
    */
   protected sealed trait ComplexKind extends AnyType.Kind
-  case object Complex extends TypeNode('Complex, AnyType) with ComplexKind {
+  case object Complex extends TypeNode('Complex, Scalar) with ComplexKind {
     type Kind = ComplexKind
   }
 
   /**
-   * For things like fn:exists, fn:empty, dfdl:contenLength
+   * For things like fn:exists, fn:empty
    */
   protected sealed trait ExistsKind extends AnyType.Kind
   case object Exists extends TypeNode('Exists, AnyType) with AnyTypeKind {
     type Kind = ExistsKind
   }
 
+  /**
+   * For things like dfdl:contentLength, dfdl:valueLength
+   */
+  protected sealed trait ScalarKind extends AnyType.Kind
+  case object Scalar
+    extends TypeNode('Scalar, AnyType, Seq(AnySimpleType, Complex))
+    with AnyTypeKind {
+    type Kind = ScalarKind
+  }
```

Comment:
I'm not familiar with the TypeNode constructor, so Steve please take a look too.

Reply:
I believe this is correct. TypeNode basically just sets up a type hierarchy. AnyType is the parent of Scalar, and it's children are simple types and complex types.

Reply:
Scalar is the opposite of Array, so I think this should be like Array, no children, no parent. 

But really we only need Exists.

I'm pretty sure we could get rid of the NodeInfo.Array easily, as it is barely used anywhere. I think it was my mistake to put Array in this taxonomy originally. I don't think it is actually used. 

The things I've seen that care if things are arrays or not aren't using NodeInfo objects or scala types like Array.Kind, they're calling PathExpression.steps.last.isArray, and I think that's what you want also. 

Actually I think you should add
```
def isArray: Boolean = steps.last.isArray
def isScalar: Boolean = !isArray
```
to PathExpression. 

Exists means "I don't care about the type" what it does care about is that the expression is a path expression i.e., where one can evaluate "does this exist". 

You want a minor variation on that where you can also ask if it's array or not. 

Dimensionality in DFDL/XSD is orthogonal to type.  I've always known that, but somehow decided a loooong time ago (like 2011) to put Array into the type hierarchy. That looks like a mistake now. 

Perhaps NodeInfo.Exists should be renamed to NodeInfo.PathExpr.

Reply:
Changed back to Exists. I can go ahead and rename Exists to PathExpr in this PR if you think we should make that change now. 

I found a isPathToOneWholeArray method in PathExpression which looks like it has the exact information needed here.